plugin/pkg/scheduler/scheduler.go

```diff
@@ -233,6 +239,99 @@ func (sched *Scheduler) preempt(preemptor *v1.Pod, scheduleErr error) (string, e
 	return node.Name, err
 }
 
+// Assume volumes and then asynchronously bind volumes if required.
+// This function modifies assumed if volume binding is required.
+func (sched *Scheduler) assumeAndBindVolumes(assumed *v1.Pod, host string) error {
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeScheduling) {
+		// TODO: can we distingush between fully bound and waiting for binding to complete?
+		// If it's waiting for binding to complete, then we can return early here instead of
+		// spawning a go routine for bindVolumes to do nothing...
+		bindingRequired, err := sched.config.VolumeBinder.Binder.AssumePodVolumes(assumed, host)
+		if err != nil {
+			sched.config.Error(assumed, err)
+			sched.config.Recorder.Eventf(assumed, v1.EventTypeWarning, "FailedScheduling", "AssumePod failed: %v", err)
+			sched.config.PodConditionUpdater.Update(assumed, &v1.PodCondition{
+				Type:    v1.PodScheduled,
+				Status:  v1.ConditionFalse,
+				Reason:  "SchedulerError",
+				Message: err.Error(),
+			})
+			return err
+		}
+		if bindingRequired {
+			if sched.config.Ecache != nil {
+				invalidPredicates := sets.NewString("NoVolumeBindConflict")
+				sched.config.Ecache.InvalidateCachedPredicateItemOfAllNodes(invalidPredicates)
+			}
+
+			sched.config.VolumeBinder.BindQueue.Add(assumed)
+
+			// bindVolumesWorker() will update the Pod object to put it back in the scheduler queue
+			return fmt.Errorf("Volume binding started, waiting for completion")
+		}
+	}
+	return nil
+}
+
+// bindVolumesWorker() processes pods queued in assumeAndBindVolumes() and tries to
+// make the API update for volume binding.
+// This function runs forever until the volume BindQueue is closed.
+func (sched *Scheduler) bindVolumesWorker() {
+	workFunc := func() bool {
+		keyObj, quit := sched.config.VolumeBinder.BindQueue.Get()
+		if quit {
+			return true
+		}
+		defer sched.config.VolumeBinder.BindQueue.Done(keyObj)
+
+		assumed, ok := keyObj.(*v1.Pod)
+		if !ok {
+			glog.V(4).Infof("Object is not a *v1.Pod")
+			return false
+		}
+
+		// TODO: add metrics
+		var reason string
+		var eventType string
+
+		glog.V(5).Infof("Trying to bind volumes for pod \"%v/%v\"", assumed.Namespace, assumed.Name)
+
+		// The Pod is always sent back to the scheduler afterwards.
+		err := sched.config.VolumeBinder.Binder.BindPodVolumes(assumed)
+		if err != nil {
+			glog.V(1).Infof("Failed to bind volumes for pod \"%v/%v\"", assumed.Namespace, assumed.Name, err)
+			reason = "VolumeBindingFailed"
+			eventType = v1.EventTypeWarning
+		} else {
+			glog.V(4).Infof("Successfully bound volumes for pod \"%v/%v\"", assumed.Namespace, assumed.Name)
+			reason = "VolumeBindingWaiting"
+			eventType = v1.EventTypeNormal
+			err = fmt.Errorf("Volume binding started, waiting for completion")
+		}
+
+		// Always fail scheduling regardless of binding success.
```

Comment:
Could you clarify this?  Are you saying that a separate informer event could cause the pod to be added to back to the queue and be processed in scheduleOne() while we are still binding in another thread?

When I look at scheduleOne() and bind() which is in a separate go routine, I see that bind() is updating the Pod outside of the main thread on error.

Reply:
> Could you clarify this? Are you saying that a separate informer event could cause the pod to be added to back to the queue and be processed in scheduleOne() while we are still binding in another thread?

Yes, it actually does cause this, but my concern is that with this change two threads compete with one another to mark the pod unschedulable. This is a bit of a concern. In case of updating the pod in bind(), scheduler main thread is already done with the pod and may not update it at the same time.

Reply:
Which part of the main thread are you concerned that it is competing with?  In `assumeAndBindVolumes`, we do not update the Pod if we queued it to the volume BindQueue.  And if we queued the pod for volume binding, the `assumeAndBindVolumes()` will return an error so that `scheduleOne()` will return immediately and not continue to `assume`.

Reply:
You are right, @msau42. I thought scheduler would continue and would update the pod in its main thread, but that's not the case. The only problem here is if the pod is updated in the middle of the volume binding process. In that case it may be retried by the scheduler, but that should be fine.

Reply:
Thanks for the clarification. I'll add some more comments to try to make it more clear.